polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp

```diff
@@ -472,6 +531,80 @@ Optional<Operation *> OperationSideEffects::conflictsWithOperationInLoop(
   return conflictsWithOperationInRegion(loop.getLoopBody());
 }
 
+//===----------------------------------------------------------------------===//
+// LoopVersionBuilder
+//===----------------------------------------------------------------------===//
+
+void LoopVersionBuilder::versionLoop() {
+  createIfOp();
+  createThenBody();
+  createElseBody();
+  replaceUsesOfLoopReturnValues();
+}
+
+void LoopVersionBuilder::replaceUsesOfLoopReturnValues() const {
+  // Replace uses of the loop return value(s) with the value(s) yielded by the
+  // if operation.
+  for (auto it : llvm::zip(loop->getResults(), ifOp->getResults()))
+    std::get<0>(it).replaceUsesWithIf(std::get<1>(it), [&](OpOperand &op) {
+      Block *useBlock = op.getOwner()->getBlock();
+      return useBlock != &getThenBlock(ifOp);
+    });
+}
+
+//===----------------------------------------------------------------------===//
+// SCFLoopVersionBuilder
+//===----------------------------------------------------------------------===//
+
+void SCFLoopVersionBuilder::createIfOp() {
+  ifOp = builder.create<scf::IfOp>(
+      loop.getLoc(), createCondition(),
+      [&](OpBuilder &b, Location loc) {
+        b.create<scf::YieldOp>(loc, loop->getResults());
+      },
+      [&](OpBuilder &b, Location loc) {
+        b.create<scf::YieldOp>(loc, loop->getResults());
+      });
+}
+
+void SCFLoopVersionBuilder::createThenBody() const {
+  loop->moveBefore(&*getThenBlock(ifOp).begin());
+}
+
+void SCFLoopVersionBuilder::createElseBody() const {
+  auto &origYield = getElseBlock(ifOp).back();
+  auto elseBodyBuilder = getIfOp().getElseBodyBuilder();
+  auto *clonedLoop = elseBodyBuilder.clone(*loop.getOperation());
+  elseBodyBuilder.create<scf::YieldOp>(loop.getLoc(), clonedLoop->getResults());
+  origYield.erase();
+}
```

Comment:
```suggestion
void SCFLoopVersionBuilder::createIfOp() {
  ifOp = builder.create<scf::IfOp>(
      loop.getLoc(), createCondition(),
      [&](OpBuilder &b, Location loc) {
        b.create<scf::YieldOp>(loc, loop->getResults());
      },
      [&](OpBuilder &b, Location loc) {
        b.create<scf::YieldOp>(loc, loop.getBody().getArguments());
      });
}

void SCFLoopVersionBuilder::createThenBody() const {
  loop->moveBefore(&*getThenBlock(ifOp).begin());
}

void SCFLoopVersionBuilder::createElseBody() const {}
```
I think this should also work